src/injected/frameCommunicators/all-frames-messenger.ts

```diff
@@ -0,0 +1,118 @@
+        private readonly pingTimeoutMilliseconds: number = 500,
+    ) {
+        this.addMessageListener(this.pingCommand, async () => {
+            await this.findResponsiveFrames();
+            return null;
+        });
+    }
+
+    public addMessageListener(
+        command: string,
+        listener: PromiseWindowCommandMessageListener,
+    ): void {
+        this.singleFrameMessenger.addMessageListener(command, listener);
+    }
+
+    public async sendMessageToWindow(
+        targetWindow: Window,
+        message: CommandMessage,
+    ): Promise<CommandMessageResponse> {
+        return this.singleFrameMessenger.sendMessageToWindow(targetWindow, message);
+    }
+
+    public async initialize(): Promise<void> {
+        await this.findResponsiveFrames();
+    }
+
+    public async sendCommandToFrames(command: string): Promise<void> {
+        if (this.responsiveFrames == null) {
+            throw new Error('AllFramesMessenger is not initialized.');
+        }
+
+        const promises: Promise<unknown>[] = this.responsiveFrames.map(frame =>
+            this.singleFrameMessenger.sendMessageToFrame(frame, {
+                command,
+            }),
+        );
+        await this.mergePromises(promises);
+    }
+
+    private async findResponsiveFrames(): Promise<void> {
+        const allIFrameElements = this.htmlElementUtils.getAllElementsByTagName(
+            'iframe',
+        ) as HTMLCollectionOf<HTMLIFrameElement>;
+
+        if (allIFrameElements.length === 0) {
+            this.responsiveFrames = [];
+            return;
+        }
+
+        const promises: Promise<unknown>[] = [];
+        for (let i = 0; i < allIFrameElements.length; i++) {
+            promises.push(
+                this.promiseFactory.timeout(
+                    this.singleFrameMessenger.sendMessageToFrame(allIFrameElements[i], {
+                        command: this.pingCommand,
+                    }),
+                    this.pingTimeoutMilliseconds,
+                ),
+            );
+        }
+
+        const results = await Promise.allSettled(promises);
+        this.responsiveFrames = [];
+        let allFramesSucceeded = true;
+
+        results.forEach((result, index) => {
+            if (result.status === 'fulfilled') {
+                this.responsiveFrames!.push(allIFrameElements[index]);
+            } else {
+                allFramesSucceeded = false;
+                const error = (result as PromiseRejectedResult).reason;
+                // We expect timeout errors if the frame message fails to send.
+                // Throw if the error is anything else.
+                if (!(error instanceof TimeoutError)) {
+                    throw error;
+                }
+            }
+        });
+
+        if (!allFramesSucceeded) {
+            this.logger.error(
+                `Some iframes could not be reached within ${this.pingTimeoutMilliseconds} milliseconds. Those frames will be ignored.`,
```

Comment:
nit: might be useful for debugging purposes to hold onto the timeout errors and include them in the logger.error call here, so that we have a stack trace/ original error message?